### DISCUSSION TITLE
Is there a way to take actions when the current user leaves a channel ?

### DISCUSSION THREAD

**Discussion by HachichaRacem** [supabase/supabase#23606] [Questions]
I am using the latest version of `Flutter 3.19.6`, `dart 3.3.4` and the `supabase_flutter` package version `2.5.2`

I have just started experimenting with the presence features and I came to the situation where I'd like to <ins>update the user's row in the database as soon as he leaves the channel</ins> whether it was intentionally (app exit) or unintentionally (lost connection, crash..etc)

What I was able to achieve so far is to make the current user "react" to other users leaving the channel or be able to inform the current of the users who have left/joined the channel

Below you can find an attempt I have made assuming that the current user id is 0.

```dart
final RealtimeChannel channel = Supabase.instance.client
    .channel('lobby', opts: const RealtimeChannelConfig(self: true));
channel.subscribe(
  (status, _) async {
    if (status == RealtimeSubscribeStatus.subscribed) {
      await channel.track(
        {'id': 0, 'hasExitedNormally': _hasExitedNormally},
      );
    }
  },
).onPresenceLeave(
  (payload) async {
    if (payload.leftPresences
            .firstWhereOrNull((element) => element.payload['id'] == 0) !=
        null) {
      await Supabase.instance.client
          .from('desktop')
          .update({'is_active': false}).eq('id', 0);
    }
  },
).onPresenceJoin((payload) {
  if (payload.newPresences
          .firstWhereOrNull((element) => element.payload['id'] == 0) !=
      null) {
    debugPrint("[JOINED]: You have successfully joined the lobby");
  }
});
```
 I am trying basically to reproduce a similar effect which was possible with firebase using the `onDisconnect` callback, is this possible using supabase ? and if yes, please guide me through it and thank you.

**Comment by GaryAustin1** [comment]
I don't use presence.  The status on .subscribe() should show connection drops and channel errors.
Also you might look at the .on('system') events and see what info might be relevant.
```
   const mySubscription2 = supabase
        .channel('mychannel2')
        .on('postgres_changes',
            { event: '*', schema: 'public', table: 'cat'}, (payload) => {
                console.log('Change received!', payload)
            })
        .subscribe((status,err)=>{
            console.log('subscribe = status',status,err);
        })
        .on('system', {}, payload => {
            console.log('system',payload)
            if (payload.extension === 'postgres_changes') {
                console.log('pc',payload);
            }
        })
```

**Comment by dshukertjr** [comment]
It seems like you want to store the real-time count of how many people are present in your app. Instead of doing it the way you are trying to do, you can just display the length of presence objects on your app.

**Comment by HachichaRacem** [comment]
Actually I am trying to kind of register a callback into my project to be triggered whenever a user leaves the channel.
It's like saying **"Hey supabase, independently from the client, if a client disconnects from the channel "X" here is their ID go and update the column Y of their row in the database to ZZZZZZ"**

This is what the official documentation of `onDisconnect` for Firebase says : 

> The OnDisconnect class is used to manage operations that will be run on the server when this client disconnects. It can be used to add or remove data based on a client's connection status. It is very useful in applications looking for 'presence' functionality. Instances of this class are obtained by calling [onDisconnect](https://firebase.google.com/docs/reference/kotlin/com/google/firebase/database/DatabaseReference#onDisconnect()) on a Firebase Database ref.

Is there any way to achieve the same behaviour?

**Reply by dshukertjr** [reply]
I understand what you are trying to do. With that being said, I am asking you to take a step back and think about what you want to achieve. Are you trying to just display the number of users that are online on your app? If that is the case, like I said, you should just use the presence feature as it is to display the online user count. If that is not the ultimate use case, then would you be able to explain what you want to ultimately achieve, and we can better assist you.

**Reply by HachichaRacem** [reply]
Thank you for replying, what I would like to achieve is to keep track of why the user left the channel/application as precisely as possible.

For example, if the user, on desktop, pressed the "X" button, it would take a moment to attempt to update a column called `exitReason` with a value for example `User exited normally` before it exists.

The case I am trying to treat, is when the user loses connection to the channel, I would like then to update the column with `User lost connection`

**Reply by Diwannist** [reply]
Hello @dshukertjr 

I want to update a database row whenever a user leaves a call on my web app. Any ideas on how I might be able to achieve this?

Specifically, I have a web app that lists live calls on the front-end. These live calls are open for visitors to join. Each live call can only have two users. Now, I need to remove a live call from the front-end when the user has disconnected. So, I want to update the database row and column to update the front-end so that users don't join a call when there is no one there.

Thank you.
